sw/device/lib/dif/dif_gpio.h

```diff
@@ -0,0 +1,327 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef SW_DEVICE_LIB_DIF_GPIO_H
+#define SW_DEVICE_LIB_DIF_GPIO_H
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/mmio.h"
+
+/**
+ * Configuration for initializing a GPIO device.
+ */
+typedef struct dif_gpio_config { mmio_region_t base_addr; } dif_gpio_config_t;
+
+/**
+ * Internal state of a GPIO device.
+ *
+ * Instances of this struct must be initialized by |dif_gpio_init()| before
+ * being passed to other functions.
+ */
+typedef struct dif_gpio { mmio_region_t base_addr; } dif_gpio_t;
+
+/**
+ * Generic return codes for the functions in this library.
+ */
+typedef enum dif_gpio_result {
+  kDifGpioResultOK,
+  kDifGpioResultInvalidArgument,
+} dif_gpio_result_t;
+
+/**
+ * Set of allowed interrupt trigger configurations.
+ */
+typedef enum dif_gpio_irq {
+  kDifGpioIrqEdgeRising,
+  kDifGpioIrqEdgeFalling,
+  kDifGpioIrqLevelLow,
+  kDifGpioIrqLevelHigh,
+  kDifGpioIrqEdgeRisingFalling,
+  kDifGpioIrqEdgeRisingLevelLow,
+  kDifGpioIrqEdgeFallingLevelHigh,
+} dif_gpio_irq_t;
```

Comment:
Neat. If we need to expand this list, we can do so later. This list seems complete enough for the moment.

Reply:
Inspired by @silvestrst [comment](https://github.com/lowRISC/opentitan/pull/1254#discussion_r371931565), thanks @silvestrst!